pkg/credentialprovider/apis/dockerauth/types.go

```diff
@@ -0,0 +1,103 @@
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DockerAuthPluginConfig is the configuration containing information about
+// each docker auth plugin. Kubelet reads this configuration from disk and enables
+// each plugin as specified by the DockerAuthPlugin type.
+type DockerAuthPluginConfig struct {
+	metav1.TypeMeta
+
+	// Plugins is a list of docker auth plugins that will be enabled by the kubelet
+	Plugins []DockerAuthPlugin
+}
+
+// DockerAuthPlugin is used by the kubelet container runtime to match the
+// image property string (from the container spec) with exec-based credential provider
+// plugins that provide container registry credentials.
+type DockerAuthPlugin struct {
+	// name is the name of the credential plugin.  It must match the name of the
+	// executable as seen by the kubelet.
+	Name string
+
+	// imageMatchers is a list of strings used to match against the image property
+	// (sometimes called "registry path") to determine which images to provide
+	// credentials for.  If one of the strings matches the image property, then the
+	// DockerAuthPlugin will be used by kubelet to provide credentials
+	// for the image pull.
+
+	// The image property of a container supports the same syntax as the docker
+	// command does, including private registries and tags. A registry path is
+	// similar to a URL, but does not contain a protocol specifier (https://).
+	//
+	// Each ImageMatcher string is a pattern which can optionally contain
+	// a port and a path, similar to the image spec.  Globs can be used in the
+	// hostname (but not the port or the path).
+	//
+	// Globs are supported as subdomains (*.k8s.io) or (k8s.*.io), and
+	// top-level-domains (k8s.*).  Matching partial subdomains is also supported
+	// (app*.k8s.io).  Each glob can only match a single subdomain segment, so
+	// *.io does not match *.k8s.io.
+	//
+	// The image property matches when it has the same number of parts as the
+	// ImageMatcher string, and each part matches.  Additionally the path of
+	// ImageMatcher must be a prefix of the target URL. If the ImageMatcher
+	// contains a port, then the port must match as well.
+	ImageMatchers []string
+
+	// ExtraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg
+}
+
+// PluginArg is used for passing extra arguments to the exec credential plugin binary.
+// Exec plugins expect the arguments to be passed in the format "key=value".
+type PluginArg struct {
+	Name  string
+	Value string
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DockerAuthPluginRequest includes the image that the kubelet requires authenication for.
+// Kubelet will pass this request object to the plugin via stdin.
+type DockerAuthPluginRequest struct {
+	metav1.TypeMeta
+
+	// image is the container image that is being pulled as part of the
+	// container registry credential request.
```

Comment:
not sure https://github.com/kubernetes/enhancements/pull/1406#discussion_r371886703 got answered ... is this the exact value in the pod container image field? would this get invoked for the configured pause image as well?

Reply:
> not sure kubernetes/enhancements#1406 (comment) got answered ... is this the exact value in the pod container image field?

My apologies, I must have missed that comment. I don't think the _exact_ image from pod spec is used since kubelet may modify it (e.g. applying defaults). But this should be the same image that is passed into the `Provide` hook of the existing built-in credential providers today.

>  would this get invoked for the configured pause image as well?

The plugin will be called in the same places we call the exisitng keyring:

https://github.com/kubernetes/kubernetes/blob/f4112710f5bbfbd9cc7a68ede0df519af30d5e40/pkg/kubelet/kuberuntime/kuberuntime_image.go#L38-L56

Today I don't think a keyring lookup happens for pause images but I'm going to scan through the kubelet code again to make sure. Either way I would think we would need to preserve behavior of the existing providers whether they provided creds for the pause image or not.